src/librbd/CreateRequest.cc

```diff
@@ -0,0 +1,633 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/CreateRequest.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "include/assert.h"
+#include "librbd/Utils.h"
+#include "common/ceph_context.h"
+#include "librbd/AioCompletion.h"
+#include "librbd/Journal.h"
+#include "librbd/MirroringWatcher.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::CreateRequest: "
+
+namespace {
+
+bool is_feature_set_inter_compatible(CephContext *cct,
+                                     uint64_t features, bool force_non_primary) {
+  if ((features & RBD_FEATURE_FAST_DIFF) != 0 &&
+      (features & RBD_FEATURE_OBJECT_MAP) == 0) {
+    lderr(cct) << "cannot use fast diff without object map" << dendl;
+    return false;
+  }
+  if ((features & RBD_FEATURE_OBJECT_MAP) != 0 &&
+      (features & RBD_FEATURE_EXCLUSIVE_LOCK) == 0) {
+    lderr(cct) << "cannot use object map without exclusive lock" << dendl;
+    return false;
+  }
+  if ((features & RBD_FEATURE_JOURNALING) != 0) {
+    if ((features & RBD_FEATURE_EXCLUSIVE_LOCK) == 0) {
+      lderr(cct) << "cannot use journaling without exclusive lock" << dendl;
+      return false;
+    }
+  } else if (force_non_primary) {
+    assert(false);
+  }
+
+  return true;
+}
+
+bool validate_layout(CephContext *cct, uint64_t size, file_layout_t &layout) {
+  if (!librbd::ObjectMap::is_compatible(layout, size)) {
+    lderr(cct) << "image size not compatible with object map" << dendl;
+    return false;
+  }
+
+  return true;
+}
+
+} // anonymous namespace
+
+namespace librbd {
+
+class ThreadPoolSingleton : public ThreadPool {
```

Comment:
Ah nice :)

I had less understanding of possible uses of ImageCtx -- in the sense, use it when there a need to open an image and things of that nature. Looking more into ImageCtx.cc, I understand it can be of used to offload more operations. Thanks! I'll do the required changes.